### ISSUE TITLE
Adding Regularizer and Optimizer in the Wrapper

### ISSUE THREAD

**Issue by paramdesai321** [GraceG1225/NeuraSectProj#27]
Add Adam, Rmsprop, SGD, adagrad for optimizers. 

Add L1, L2, Dropout for Regularizers.

**Comment by paramdesai321** [comment]
@GraceG1225 it is not viable to merge MLP(#31) to main because as you can see in #31 there are envs pushes in MLP(related issue that Gev is working on resolving #22). So until he gets rid of those envs and cache files from MLP we shuold not transfer those to main. Long story short I created a branch called **Optimizers-and-regularizers** which you should use to work on this issue. 

To work on this branch from your computer(local machine):

1. Make sure your branch MLP is up to date with the remote branch: 
In the branch MLP on your machine do:
`git status`
this should return 
`Your branch is up to date with 'origin/MLP'.`
if not then let me know ASAP. 
2. Switch to the branch  **Optimizers-and-regularizers**
`git switch -c Optimizers-and-regularizers`
this will take you to the new branch. 

In this new branch you shuold have the same setup as branch **MLP**

**Comment by paramdesai321** [comment]
Simple dropout example:
``` python
import tensorflow as tf
from tensorflow.keras import layers, models

model = models.Sequential([
    layers.Dense(128, activation='relu', input_shape=(784,)),
    layers.Dropout(0.3),  # Drops 30% of neurons during training
    
    layers.Dense(64, activation='relu'),
    layers.Dropout(0.2),  # Drops 20% of neurons during training
    
    layers.Dense(10, activation='softmax')
])

```

**Comment by paramdesai321** [comment]
Dropout can be made a sub-class of layer and we can customized options like the following: 

``` python
class Dropout(Layer):
    def __init__(self, rate=0.2, noise_shape=None, seed=None):
        self.rate = rate 
        self. ...

    def per_layer_fixed_rate(self,rate=0.2):
        ...
    def per_layer_choosed_rate(self,rate):
        ...
    def customized_rate_increasing(self, min_rate=0.2,max_rate=0.8,alpha=0.2):
        ...
    def customized_rate_decreasing(self, min_rate=0.2,max_rate=0.8,alpha=0.2):
        ...

```

**Comment by paramdesai321** [comment]
Optimizer can be put in with the functions(like compile, summary, evaluate) in class MLP. You can do something similar or use that entirely if that's correct. Below I have pasted my **get_optimizer** function and it's use in **compile** function

``` python        
    def _get_optimizer(self, optimizer):
    if isinstance(optimizer, str):
        opt = optimizer.lower()

        if opt == "adam":
            return tf.keras.optimizers.Adam()
        elif opt == "sgd":
            return tf.keras.optimizers.SGD()
        elif opt == "rmsprop":
            return tf.keras.optimizers.RMSprop()
        elif opt == "adagrad":
            return tf.keras.optimizers.Adagrad()
        else:
            raise ValueError(f"Unsupported optimizer: {optimizer}")

    return optimizer  # already an optimizer object

    def compile(self, optimizer="sgd", loss=None, metrics=None):

        if self._compiled:
            return # compile() can no longer be called multiple times

    # build model (tf_build sets self.model)
        self.tf_build()

    # choose correct default loss
        if loss is None:
            if getattr(self,"output_layer",None) is not None and getattr(self.output_layer,"activation","")=="softmax":
                loss = "categorical_crossentropy"
            else:
                loss = "mse"
        opt = self._get_optimizer(optimizer)
        self.model.compile(optimizer=opt, loss=loss, metrics=metrics or [])
        self._compiled = True
```

**Comment by GraceG1225** [comment]
Method for regularizers. 6/9/26
```python
class Layer:
    def __init__(self, input_shape=None, neurons=2, activation="relu", layer_type="Dense", regularizer = None,
                 regularization_rate=0.00, dropout_rate=0.00):
        self.input_shape = input_shape
        self.neurons = neurons
        self.activation = activation
        self.layer_type = layer_type
        self.regularizer = regularizer
        self.regularization_rate = regularization_rate
        self.dropout_rate = dropout_rate
        self.prev = None
        self.next = None`
    
    def build_regularizer(self):
        # share to layer and conv to return keras regularizer obj or nothing
        if self.regularizer == "l1":
            return tf.keras.regularizers.L1(self.regularization_rate)
        elif self.regularizer == "l2":
            return tf.keras.regularizers.L2(self.regularization_rate)
        return None

    def _get_conv_layer(self):
        # will call inherited regularizer
        reg = self._build_regularizer()

```

**Comment by GraceG1225** [comment]
Please see `class Dropout` and `apply_dropout()` in MLPClass.ipynb

**Comment by paramdesai321** [comment]
Everything seems good. Except that the training results are very volatile with different Dropout modes(especially the ones with variable rate). I'll investigate if those are natural or if there is a bug.

**Comment by GraceG1225** [comment]
Modified tf_build function in MLPClass.ipynb.

Previously was:
`def tf_build(self):`
`    if self.model.layers:`
`        return self.model # in case of already built`

Changed to be:
`def tf_build(self):`
`    if self.model.layers and self._built:`
`        return self.model # in case of already built`

The issue was as follows: compile() calls tf_build(), which builds and caches the Sequential model before dropout nodes have been inserted into the linked list. When tf_build() is called again later, self.model.layers is non-empty from that first build, so it returns the cached pre-dropout model. By changing the condition, tf_build() only skips rebuilding when both are true. The Sequential model exists **and** the linked list hasn't been modified since the model was last built. Apply_dropout() resets _built = False, so the next call to tf_build() is forced to re-traverse the linked list and include the dropout nodes.
